### PR TITLE
[FLINK-30579][hive] Introducing cofigurable option to enable hive native function

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

*Introducing query-level configurable options `table.exec.hive.native-agg-function.enabled` to support enable native hive aggregate function, native implementation use hash-agg strategy that can improve the aggregate performance. However, due to flink current ability, we can't assign the native agg function behavior with hive udaf, so if the performance isn't important, you shouldn't enable this option.*

## Brief change log

  - *Introducing query-level configurable options `table.exec.hive.native-agg-function.enabled` to support enable native hive aggregate function*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added integration tests in HiveDialectAggITCase*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes /no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by lsyldliu** [comment]
@flinkbot run azure

**Comment by lsyldliu** [comment]
cc @luoyuxia

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSumAggFunction.java [hunk 1]
```diff
@@ -125,10 +126,16 @@ private DataType initResultType(DataType argsType) {
...
+                throw new TableException(
+                        String.format(
+                                "Native hive sum aggregate function does not support type: '%s'. Please set option '%s' to false.",
+                                argsType.getLogicalType().getTypeRoot(),
+                                TABLE_EXEC_HIVE_NATIVE_AGG_FUNCTION_ENABLED.key()));
             default:
                 throw new TableException(
                         String.format(
-                                "Sum aggregate function does not support type: '%s'. Please re-check the data type.",
+                                "Native hive sum aggregate function does not support type: '%s'. Please re-check the data type.",
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSumAggFunction.java:138]: Combine this with `case TIMESTAMP_WITHOUT_TIME_ZONE` ?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectAggITCase.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+                .addRow(new Object[] {1, 1})
+                .addRow(new Object[] {2, 2})
+                .addRow(new Object[] {3, 3})
+                .addRow(new Object[] {4, 4})
+                .addRow(new Object[] {5, 5})
+                .commit();
+    }
+
+    @Test
+    public void testSumAggFunctionPlan() {
```
Comment (luoyuxia) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveDialectAggITCase.java:75]: add a test to validate the fall back can work?

File: flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.flink.table.module.Module;
+
+import java.util.Map;
+
+/**
+ * An interface used to create {@link Module} or {@link Catalog}.
+ *
+ * @param <R> type of the return value
+ */
+@PublicEvolving
```
Comment (luoyuxia) [inline flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java:33]: is it must a PublicEvolving?

**Review by luoyuxia — COMMENTED** [review]
@lsyldliu Thanks for contribution. I have a overview about Hive Module & Test part and left minor comments. @fsk119 may help review the HiveServer2Endpoint part.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSumAggFunction.java (hunk 1 above)
Comment (lsyldliu) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSumAggFunction.java:138]: For the `TIMESTAMP_WITHOUT_TIME_ZONE ` type, hive sum udaf supports it, but flink can't support, but for default case type, hive  sum udaf and flink both can't support it, so I think should distinguish these two case, the exception shouldn't be same.

File: flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java (hunk 3 above)
Comment (lsyldliu) [reply flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java:33]: The `SessionEnvironment` is PublicEvovling, so here annotate it.

**Comment by lsyldliu** [comment]
@luoyuxia Thanks for review, I've address your comments.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSumAggFunction.java (hunk 1 above)
Comment (luoyuxia) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveSumAggFunction.java:138]: May be we can have a exception message following Hive like:
`"Only numeric or string type arguments are accepted but "
              + parameters[0].getTypeName() + " is passed."`

File: flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+public interface FunctionCreator<R> {
+
+    /**
+     * Applies this function to the given arguments.
+     *
+     * @return the function result
```
Comment (fsk119) [inline flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java:39]: Add a description of the input parameters as well.

File: flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    R create(Map<String, String> options, ReadableConfig configuration, ClassLoader classLoader);
```
Comment (fsk119) [inline flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java:41]: I think we should not let the SqlGatewayService determine the options. It's the `Endpoint`'s responsibility to determine the options.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/endpoint/hive/HiveServer2Endpoint.java [hunk 6]
```diff
@@ -310,7 +311,11 @@ public TOpenSessionResp OpenSession(TOpenSessionReq tOpenSessionReq) throws TExc
             // all the alive PersistenceManager in the ObjectStore, which may get error like
             // "Persistence Manager has been closed" in the later connection.
             hiveCatalog.open();
-            Module hiveModule = new HiveModule();
+            // create hive module lazily
+            FunctionCreator<Module> hiveModuleCreator =
+                    (options, readableConfig, classLoader) ->
+                            FactoryUtil.createModule(
```
Comment (fsk119) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/endpoint/hive/HiveServer2Endpoint.java:317]: Notice: Factory#createModule still uses the AppClassloader to find the Factory. Please take at look at the `FactoryUtil#createModule` Line 486

**Review by fsk119 — COMMENTED** [review]
Thanks for your contribution. I left some comments.

File: flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java (hunk 3 above)
Comment (lsyldliu) [reply flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java:33]: Sorry, I don't get your point, can you explain more?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/endpoint/hive/HiveServer2Endpoint.java (hunk 6 above)
Comment (lsyldliu) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/endpoint/hive/HiveServer2Endpoint.java:315]: It uses the classloader in Context, and so is user classloader.

**Review by fsk119 — COMMENTED** [review]
LGTM. Please also add a test in the HiveServer2EndpointStatementITCase.

**Comment by lsyldliu** [comment]
@fsk119 Thanks for reviewing, the test has been added.

**Review by fsk119 — APPROVED** [review]
LGTM

**Pull request closed without merging**